components/feature/session/src/main/java/mozilla/components/feature/session/ThumbnailsFeature.kt

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.session
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import mozilla.components.browser.session.SelectionAwareSessionObserver
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.concept.engine.EngineView
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.ktx.android.content.isOSOnLowMemory
+
+/**
+ * Feature implementation for automatically taking thumbnails of sites.
+ */
+class ThumbnailsFeature(
+    private val context: Context,
+    private val engineView: EngineView,
+    sessionManager: SessionManager
+) : LifecycleAwareFeature {
+
+    private val observer = ThumbnailsFeatureRequestObserver(sessionManager)
+
+    /**
+     * Starts observing the selected session to listen for when a session finish loading.
+     */
+    override fun start() {
+        observer.observeSelected()
+    }
+
+    /**
+     * Stops observing the selected session.
+     */
+    override fun stop() {
+        observer.stop()
+    }
+
+    internal inner class ThumbnailsFeatureRequestObserver(
+        sessionManager: SessionManager
+    ) : SelectionAwareSessionObserver(sessionManager) {
+
+        override fun onProgress(session: Session, progress: Int) {
+            if (progress == PROGRESS_STOP) {
+                requestPrintScreen(session)
+            }
```

Comment:
Instead of "onProgress" you may want to look at `onLoadingStateChanged`.

Interestingly Fennec delays taking the screenshot (500ms) after receiving the state change:
https://searchfox.org/mozilla-central/source/mobile/android/base/java/org/mozilla/gecko/Tab.java#718

Reply:
What would be the main difference between them? They are getting called in very similar places
![image](https://user-images.githubusercontent.com/773158/54222854-9d9f7100-44cc-11e9-9642-71fe6ad5ae82.png)

https://github.com/mozilla-mobile/android-components/blob/e5cefefbcecf898c07fd149faea20f94015a168d/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt#L367

Reply:
This is an implementation detail though. There's no guarantee that this will always be the case. For example this is different in `browser-engine-system`. Also there's no guarantee that the loading state will always hit 100% (but it will stop).

Reply:
👍